servers/rendering/renderer_rd/api_context_rd.h

```diff
@@ -0,0 +1,77 @@
+/*                             GODOT ENGINE                               */
+/*                        https://godotengine.org                         */
+/**************************************************************************/
+/* Copyright (c) 2014-present Godot Engine contributors (see AUTHORS.md). */
+/* Copyright (c) 2007-2014 Juan Linietsky, Ariel Manzur.                  */
+/*                                                                        */
+/* Permission is hereby granted, free of charge, to any person obtaining  */
+/* a copy of this software and associated documentation files (the        */
+/* "Software"), to deal in the Software without restriction, including    */
+/* without limitation the rights to use, copy, modify, merge, publish,    */
+/* distribute, sublicense, and/or sell copies of the Software, and to     */
+/* permit persons to whom the Software is furnished to do so, subject to  */
+/* the following conditions:                                              */
+/*                                                                        */
+/* The above copyright notice and this permission notice shall be         */
+/* included in all copies or substantial portions of the Software.        */
+/*                                                                        */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,        */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF     */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. */
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY   */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,   */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE      */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
+/**************************************************************************/
+
+#ifndef API_CONTEXT_RD_H
+#define API_CONTEXT_RD_H
+
+#include "servers/rendering/rendering_device.h"
+#include "servers/rendering/rendering_device_driver.h"
+
+class ApiContextRD {
+public:
+	virtual RenderingDevice::Capabilities get_device_capabilities() const = 0;
+	virtual const RDD::MultiviewCapabilities &get_multiview_capabilities() const = 0;
+
+	virtual int get_swapchain_image_count() const = 0;
+
+	virtual int window_get_width(DisplayServer::WindowID p_window = 0) = 0;
+	virtual int window_get_height(DisplayServer::WindowID p_window = 0) = 0;
+	virtual bool window_is_valid_swapchain(DisplayServer::WindowID p_window = 0) = 0;
+	virtual RDD::RenderPassID window_get_render_pass(DisplayServer::WindowID p_window = 0) = 0;
+	virtual RDD::FramebufferID window_get_framebuffer(DisplayServer::WindowID p_window = 0) = 0;
+
+	virtual RID local_device_create() = 0;
+	virtual void local_device_push_command_buffers(RID p_local_device, const RDD::CommandBufferID *p_buffers, int p_count) = 0;
```

Comment:
This will probably need to be changed for the acyclic graph, exposing queues and synchronization primitives.

Reply:
For now, I've chosen to alter the design of the context class as little as possible to keep the scope of the work under control for this first big iteration, but I won't deny future maintainers will need to make this more flexible (probably exposing from the driver).